crates/turborepo-lib/src/shim.rs

```diff
@@ -176,50 +178,234 @@ struct PackageJson {
+            .lines()
+            .next()
+            .expect("Failed to read version from version.txt")
+    }
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct LocalTurboState {
+    bin_path: PathBuf,
+    version: String,
+}
+
+impl LocalTurboState {
+    // Hoisted strategy:
+    // - `npm install`
+    // - `yarn`
+    // - `yarn install --flat`
+    // - berry (nodeLinker: "node-modules")
+    //
+    // This also supports people directly depending upon the platform version.
+    fn generate_hoisted_path(root_path: &Path) -> Option<PathBuf> {
+        Some(root_path.join("node_modules"))
+    }
+
+    // Nested strategy:
+    // - `npm install --install-strategy=shallow` (`npm install --global-style`)
+    // - `npm install --install-strategy=nested` (`npm install --legacy-bundling`)
+    // - berry (nodeLinker: "pnpm")
+    fn generate_nested_path(root_path: &Path) -> Option<PathBuf> {
+        Some(
+            root_path
+                .join("node_modules")
+                .join("turbo")
+                .join("node_modules"),
+        )
+    }
 
-        let local_turbo_package_path = repo_root
-            .join("node_modules")
-            .join("turbo")
-            .join("package.json");
+    // Linked strategy:
+    // - `pnpm install`
+    // - `npm install --install-strategy=linked`
+    fn generate_linked_path(root_path: &Path) -> Option<PathBuf> {
+        fs_canonicalize(root_path.join("node_modules").join("turbo").join("..")).ok()
+    }
+
+    // The unplugged directory doesn't have a fixed path.
+    fn get_unplugged_base_path(root_path: &Path) -> PathBuf {
+        let yarn_rc_filename =
+            env::var_os("YARN_RC_FILENAME").unwrap_or_else(|| OsString::from(".yarnrc.yml"));
+        let yarn_rc_filepath = root_path.join(yarn_rc_filename);
+
+        let yarn_rc_yaml_string = fs::read_to_string(yarn_rc_filepath).unwrap_or_default();
+        let yarn_rc: YarnRc = serde_yaml::from_str(&yarn_rc_yaml_string).unwrap_or_default();
+
+        root_path.join(yarn_rc.pnp_unplugged_folder)
+    }
 
-        let package_json: PackageJson =
-            serde_json::from_reader(File::open(local_turbo_package_path).ok()?).ok()?;
+    // Unplugged strategy:
+    // - berry 2.1+
+    fn generate_unplugged_path(root_path: &Path) -> Option<PathBuf> {
+        let platform_package_name = TurboState::platform_package_name();
+        let unplugged_base_path = Self::get_unplugged_base_path(root_path);
+
+        unplugged_base_path
+            .read_dir()
+            .ok()
+            .and_then(|mut read_dir| {
+                // berry includes additional metadata in the filename.
+                // We actually have to find the platform package.
+                read_dir.find_map(|item| match item {
+                    Ok(entry) => {
+                        let file_name = entry.file_name();
+                        if file_name
+                            .to_string_lossy()
```

Comment:
We should probably decide on a strategy for handling these types of issues.

Reply:
It doesn't actually matter for this case which is why I chose this serialization:
- The upstream producers of these file paths only generate Unicode.
- If you try to break it by putting invalid Unicode in the middle of the prefix you get the replacement character and we don't match.
- If you try to break it by putting invalid Unicode after the prefix you get the replacement character and we actually do the right thing and proceed.